pymc/aesaraf.py

```diff
@@ -1015,3 +1019,144 @@ def compile_pymc(
+    considered volatile, and as such, they wont be included as inputs into the compiled function.
+    This means that if they have values stored in the posterior, these values will be ignored
+    and new values will be computed (in the case of deterministics and potentials) or sampled
+    (in the case of random variables).
+
+    This function also enables a way to impute values for any variable in the computational
+    graph that produces the desired outputs: the ``givens_dict``. This dictionary can be used
+    to set the ``givens`` argument of the aesara function compilation. This will essentially
+    replace a node in the computational graph with any other expression that has the same
+    type as the desired node. Passing variables in the givens_dict is considered an intervention
+    that might lead to different variable values from those that could have been seen during
+    inference, as such, **any variable that is passed in the ``givens_dict`` will be considered
+    volatile**.
+
+    Parameters
+    ----------
+    outputs : List[aesara.graph.basic.Variable]
+        The list of variables that will be returned by the compiled function
+    vars_in_trace : List[aesara.graph.basic.Variable]
+        The list of variables that are assumed to have values stored in the trace
+    givens_dict : Optional[Dict[aesara.graph.basic.Variable, Any]]
+        A dictionary that maps tensor variables to the values that should be used to replace them
+        in the compiled function. The types of the key and value should match or an error will be
+        raised during compilation.
+    """
+    if givens_dict is None:
+        givens_dict = {}
+    # We need a function graph to walk the clients and propagate the volatile property
+    fg = aesara.graph.FunctionGraph(outputs=outputs, clone=False)
+
+    # Walk the graph from inputs to outputs and tag the volatile variables
+    nodes = general_toposort(fg.outputs, deps=lambda x: x.get_parents())
+    givens = []
+    registered_givens = set()
+    volatile_nodes = set()
+    for node in nodes:
+        if (
+            isinstance(node, aesara.compile.SharedVariable)
+            or node in fg.outputs
+            or node in givens_dict
+        ):
+            volatile_nodes.add(node)
+        is_volatile = node in volatile_nodes
+
+        # Populate the givens list if the node is in the givens_dict
+        if node in givens_dict and node not in registered_givens:
+            value = givens_dict[node]
+            if not isinstance(value, (Apply, Variable)):
+                # If the value is not a tensor, it is wrapped as a Constant with the node's name
+                value = at.constant(
+                    givens_dict[node],
+                    dtype=node.dtype,
+                    name=node.name,
+                )
+            givens.append((node, value))
+            registered_givens.add(node)
```

Comment:
This does not seem like it needs to be in this loop, it has nothing to do with volatility inference

Reply:
Yeah, I'll take it out